### ISSUE TITLE
Type deduction for records

### ISSUE THREAD

**Issue by julianhyde** [julianhyde/smlj#139]
Improve type deduction for records. For example,
```
fun hasJob e job =
  e.job = job;
```
throws
```
java.lang.ClassCastException: net.hydromatic.morel.type.TypeVar cannot be cast to net.hydromatic.morel.type.RecordLikeType
```
and has to be changed to
```
fun hasJob (e: {name: string, job: string, empno: int}) job =
  e.job = job;
```
in order to compile. Ideally we would infer record types (or constraints on record types) but at a minimum we should give a good error, not throw `ClassCastException`.

**Comment by gfrmin** [comment]
@julianhyde, this is the "scary program" you set me by email. It's the same gap as #139, so I'll put it here rather than open a duplicate.

A function whose parameter is constrained only by the fields it selects ought to infer a flex (row-polymorphic) record type, with no annotation. At present the record type stays an unresolved type variable and resolution gives up. That blocks the one pattern a dbt-style transform library most needs: a named function that filters or projects rows by field name.

Fails today, next to your `hasJob`:

```
fun singleUnitItems items = from i in items where i.units = 1;
> stdIn:1.51 Error: unresolved flex record (can't tell what fields there are besides #units)
```

Your case. Desired: infer `{job:string, ...} -> int`, accepted on records that also carry an `ename` field, no annotation:

```
fun f emps = from e in emps where e.job = "MANAGER" compute count over ();
```

Works today only once you spell the whole record out:

```
fun singleUnitItems (items: {units:int} list) = from i in items where i.units = 1;
> val singleUnitItems = fn : {units:int} list -> {units:int} list
```

Raised in `TypeResolver.checkNoUnresolvedFieldRefs`: the `variableConsumer` lambda at `TypeResolver.java:290-296` (applied selector `#units i`); the bare-selector path is `:783-788`. Adjacent but distinct: #375 (`yield` not exposing fields), #312 (add-a-field syntax), #299 (list/bag element ambiguity).

I take your point that the bar is an elegant solution in the inference, not a hack bolted to the side, and that the alternative is people quietly reaching for Jinja. Happy to prototype against this and your `count` case as the tests that must pass. The fix is all in inference, so there's no new language surface to regret later.

**Comment by gfrmin** [comment]
As promised in June: the design, before any code, since the bar you set was an elegant solution in the inference rather than a hack bolted to the side. It is long because the alternatives needed eliminating on paper; the recommendation itself is short.

First, housekeeping: the `ClassCastException` in the issue text is gone. At HEAD, `fun hasJob e job = e.job = job` fails with the positioned `unresolved flex record (can't tell what fields there are besides #job)`. So the minimum ask is met, and what remains is the real one: inferring the type instead of apologising for not knowing it.

## What the machinery does today

Three facts about the current implementation shape the whole design space.

**Records cannot be partially known.** The inferencer is a textbook first-order Martelli-Montanari solver (`MartelliUnifier`); its terms are variables and operator-applied sequences, and a record type's operator string carries its complete sorted label set: `{empno:int, job:string}` becomes the operator `record:empno:job` (`recordLabel`, `TypeResolver.java:3020`). Records with different field sets are different operators and can never unify. There is nowhere in the term language to write "these fields, and some others I have not met yet".

**Field selection consumes record types but never produces them.** `e.job` does not emit an equation. `deduceRecordSelectorType` (`TypeResolver.java:2882`) registers a passive callback on the receiver's variable: if something else resolves that variable to a concrete record, project the field. For a function parameter there is no something else, the variable stays free, and the post-pass (`checkNoUnresolvedFieldRefs`, `:356`) gives up. Information flows one way, which is precisely the wrong way for `hasJob`.

**Your `count` example has a second, independent gap.** `from e in emps` leaves the orderedness of `emps` open: a deferred disjunctive constraint (`mayBeBagOrList`, `:630`), riding the same constraint queue as overload resolution, says the collection is `list(v)` or `bag(v)`. If neither candidate wins by the end of the declaration, the constraint is silently dropped (`MartelliUnifier.java:148`) and the variable generalises to a bare `'a`; the Resolver then dies with a raw `IllegalArgumentException: not a collection type: 'a` (`Type.java:158`; this is #299's family). So `fun f emps = from e in emps where e.job = "MANAGER" compute count over ()` needs two answers, not one: what fields `e` has, and what the silently dropped disjunction should have become.

## Recommendation: kinded type variables (Ohori)

The observation that makes me comfortable recommending this: Morel has already built two thirds of SML#'s kinded unification without calling it that. An overloaded name is a type variable constrained to range over an enumerated set of types. A collection of unknown orderedness is a type variable constrained to be list-of or bag-of. Both are deferred, merged, and discharged by the same constraint queue. What is missing is the third kind in Ohori's calculus (A Polymorphic Record Calculus and Its Compilation, TOPLAS 1995, the system under SML#): the record kind, a variable constrained to have *at least* these fields. Add it and record inference falls out. This is not a third side channel bolted beside the two you already tolerate; it is the theory the existing two were groping towards, with one lifecycle for all three: constrain, merge, discharge, and (the new part) *residuate* instead of silently dropping. And the calculus descends from Ohori and Buneman's work on database programming languages (Machiavelli), so the pedigree could hardly be more on the nose.

Mechanism, tied to the code as it stands:

* A conjunctive constraint `FieldsOf(v, {job: t, ...})` joins today's disjunctive one (`Unifier.Constraint`, discharged in `MartelliUnifier.Work.subConstraint`, `:304`). `e.job` emits `FieldsOf(typeof e, {job: a})` in place of the passive callback; `a` is an ordinary variable usable downstream, so `= "MANAGER"` refines the requirement to `{job: string}` with no extra machinery. Bare `#job` becomes a first-class value of type `'a#{job:'b} -> 'b`, so `map #job emps` typechecks too.
* **Merge**: when two constrained variables are unified, requirement maps union and common labels' types are equated (Ohori's kind lub). The substitution scan that already retargets constraints runs on every binding, including variable-to-variable, so this needs no new unifier hooks.
* **Discharge**: when the variable resolves to a concrete record term, subset test plus per-field unification. A missing field reports the field name and the requiring site. A progressive record (the `z$dummy` sentinel behind foreign schemas) defers instead of failing, and the existing `FieldExpander` retry loop discovers the field; external catalogues keep working unchanged.
* **Residuate**: requirements alive at the end of a top-level declaration generalise into a kinded type variable rather than erroring. Instantiation re-emits the requirement on the fresh variable, closing the loop: applying the function to a record lacking the field fails with "required field 'job' is missing"; applying it to a *wider* record succeeds, because discharge is a subset test. Width tolerance with no subtyping anywhere, which is what lets Ohori prove principal typings for the record fragment; the overload fragment keeps today's commit-when-one-candidate-remains heuristic, no better and no worse.
* **Zero new grammar.** Kinds print (SML#'s notation, `'a#{job:string}`) but nothing new parses. `...` stays pattern-only. The language grows no surface at all in the first two stages, which I believe is the answer to your "does Morel still feel small" test.

What the acceptance programs infer:

```sml
fun hasJob e job = e.job = job;
> val hasJob = fn : 'a#{job:'b} -> 'b -> bool

fun singleUnitItems items = from i in items where i.units = 1;
> val singleUnitItems = fn : 'a#{units:int} bag -> 'a bag

fun f emps = from e in emps where e.job = "MANAGER" compute count over ();
> val f = fn : 'a#{job:string} bag -> int
```

The `bag` in the last two is the orderedness decision, and I want your ruling on it. Two options:

* **(A) Residuate the disjunction too.** The same end-of-declaration policy that turns a field requirement into a record kind turns the surviving list-or-bag constraint into an overload kind on the collection constructor: `f` works on lists and bags alike, `singleUnitItems` maps list to list and bag to bag. This is quantification over a constructor drawn from a two-element set, a modest step beyond plain HM but one your constraint queue already takes at every overloaded application; and it retires the `not a collection type: 'a` crash class by construction.
* **(B) Default surviving orderedness to `bag`.** Small, SQL-flavoured (a query source of unknown order is a bag), decouples #139 from the orderedness question entirely. Cost: `f [r1, r2]` on a list literal needs a conversion, which will surprise people.

I lean (A) for the theory and (B) for the diff; happy either way, but one of them has to ship inside this feature or your `count` program still fails on its second gap.

## The alternative, for the record

The other principled design is Rémy/Wand row variables: record terms grow a tail slot, `e.job` emits the equation `typeof e = {job: a; r}`, and the decompose rule learns to unify records label-selectively, spilling each side's surplus fields into the other's tail. It is the textbook solution and it would work here; Morel needs no lacks predicates (no record extension operator, so duplicate labels cannot arise), which removes the usual sting. I am not recommending it because it modifies the core unification algorithm itself (the conflict rule becomes theory-aware), perturbs every consumer of record terms with the tail slot, and does nothing for the orderedness gap, whereas kinds ride infrastructure you already maintain. The one future in which rows age better: if #312 (syntax to add a field to a record) ever grows into polymorphic record extension, rows express record *construction* polymorphism and kinds do not. If you see that on the roadmap, say so now and I will re-argue this section.

## The cost neither design escapes

Runtime records are label-stripped positional lists, and `Core.RecordSelector` carries a compile-time integer slot (`Core.java:819`) compiled to `Codes.nth(slot)` (`Compiler.java:550`). A function body that is polymorphic over its receiver's fields has no slot. Ohori's paper solves this too (it is the "and Its Compilation" half): evidence passing, one hidden `int` parameter per generalised selection, supplied at each instantiation site where the concrete record type is known. Two mitigations make this cheaper in Morel than in a general-purpose SML: on the relational path the inliner folds the function into the query and the evidence constant-folds away before Calcite ever sees it; and records built, passed, or compared whole need no evidence at all, only selection does. But it is the single largest work item in the feature, it is shared by rows and kinds alike, and it must eventually be done twice (#385), so it deserves a decision at design time rather than discovery at review time.

## Staging

1. **Constraint infrastructure, no new power.** `FieldsOf` with merge and discharge; selectors move off the passive callback (the callback's projection becomes the discharge path, so behaviour on currently-legal programs should be unchanged, and the existing suite is the referee). Immediate dividend: the flex-record error can list the full accumulated requirement (`e must have at least {job:string, units:int}`) instead of the first field it happened to meet.
2. **Kinded generalisation and instantiation, plus the (A)/(B) decision.** The three programs above infer and print. Kinded variables are rejected where they cannot yet mean anything (explicit user type variables, datatype declarations).
3. **Evidence passing.** The three programs run.
4. **Flex patterns.** `fun f {a=x, ...} = x` stops being a parse-then-reject and means what it says (a requirement on the pattern's variable). Note this is a divergence: SML'97 insists the ellipsis be resolved from context. I think generalising is the consistent choice, since the pattern is the dual of the selector, but it deviates from the Definition and deserves your explicit blessing.

Each stage is a separate PR against the tests that must pass: your `hasJob`, your `count` program, and my `singleUnitItems`. I would prototype stages 1 and 2 on a branch first and bring the results here before opening anything.

Open questions, beyond (A)/(B): whether the printed notation should be SML#'s `'a#{job:string}` or an anonymised `{job:string, ...}` when the variable occurs once (I lean the former; it survives the variable occurring twice, as in `singleUnitItems`); whether numeric labels may generalise (`fun fst p = #1 p` as `'a#{1:'b} -> 'b`, which Ohori permits and tuple-lovers may find alarming); and a note rather than a question: Morel currently has no value restriction anywhere, and while kinds add no new unsoundness (there are no ref cells to exploit), whoever eventually adds effects inherits the reckoning for type variables and kinded variables together.
